sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,242 @@
+from sympy.physics.mechanics.body import Body
+from sympy.physics.vector import Vector, dynamicsymbols
+from abc import ABC, abstractmethod
+
+__all__ = ['Joint', 'PinJoint']
+
+
+class Joint(ABC):
+    """Abstract Base class for all specific joints.
+
+    A Joint connects two bodies (a parent and child) by adding different degrees
+    of freedom to child with respect to parent.
+    This is the base class for all specific joints and holds all common methods
+    acting as an interface for all joints. Custom joint can be created by
+    inheriting Joint class and defining all abstract functions.
+
+    Parameters
+    ==========
+
+    name : string
+        The Joint's name which makes it unique.
+    parent : Body
+        The parent body of joint.
+    child : Body
+        The child body of joint.
+    dist : Vector
+        The vector distance between parent's masscenter and child's masscenter.
+    coordinates: List of dynamicsymbols/dynamicsymbol, optional
+        Coordinates of joint.
+    speeds : List of dynamicsymbols/dynamicsymbol, optional
+        Speeds of joint.
+    parent_joint_pos : Vector, optional
+        Defines the joint's point where the parent will be connected to child.
+        Default value is masscenter of Parent Body.
+    child_joint_pos : Vector, optional
+        Defines the joint's point where the child will be connected to parent.
+        Default value is masscenter of Child Body.
+    parent_axis : Vector, optional
+        Axis of parent frame which would be be aligned with child's
+        axis. Default is x axis in parent's frame.
+    child_axis : Vector, optional
+        Axis of Child frame which would be be aligned with parent's
+        axis. Default is x axis in child's frame.
+
+    """
+
+    def __init__(self, name, parent, child, dist, coordinates=None, speeds=None, parent_joint_pos=None,
+        child_joint_pos=None, parent_axis = None):
+
+        if not isinstance(name, str):
+            raise TypeError('Supply a valid name.')
+        self._name = name
+
+        if not isinstance(parent, Body):
+            raise TypeError('Parent must be an instance of Body.')
+        self._parent = parent
+
+        if not isinstance(child, Body):
+            raise TypeError('Parent must be an instance of Body.')
```

Comment:
```suggestion
            raise TypeError('Child must be an instance of Body.')
```
I also noticed some other typos.